app/assets/javascripts/nbp_wallet.coffee.erb

```diff
@@ -0,0 +1,28 @@
+TEMPLATE_VALIDITY_PERIOD = <%= Enmeshed::ConnectorApi::TEMPLATE_VALIDITY_PERIOD %>
+
+template_validity = TEMPLATE_VALIDITY_PERIOD
+
+checkStatus = ->
+  if window.location.pathname == "/users/nbp_wallet/connect"
+    $.ajax(
+      url: "/users/nbp_wallet/relationship_status"
+      method: "POST"
+      success: (data, textStatus, xhr) ->
+        if data.status == 'ready'
+          window.location = "/users/nbp_wallet/finalize"
+    )
+
+countdownValidity = ->
+  if template_validity > 0
+    template_validity -= 1
+    if template_validity == 0
+      window.location.reload()
```

Comment:
The automatic reloading might invalidate a current connection request still open:

1. Get the QR code with 120 seconds validity
2. Open the app, scan it, enter your details
3. Let the QR code expire (or be too slow)
4. Send the connection request in the app.

At least, the app seems to handle this case now, but still it just hides the form without any message (and CodeHarbor doesn't show one, either).

Reply:
If this error happened (and one tries a second time to connect), the attributes are shown twice in the Enmeshed app for the given contact.

<details><summary>Details</summary>
<p>

![IMG_1105](https://github.com/openHPI/codeharbor/assets/7300329/b869a7e5-d60b-4e07-a228-f1aea6c625c6)
![IMG_1106](https://github.com/openHPI/codeharbor/assets/7300329/aead5438-30b0-47df-b1cc-e04f3bfad930)

</p>
</details>

Reply:
We could make the QR codes live longer (60 minutes for example) to avoid this but then the following becomes a problem:
1. user scans QR code A (but does not enter data yet)
2. user clicks refresh QR code (at which point the relationship_template_id in the session gets replaced which means we cannot associate QR code A with the user anymore)
3. the user submits their data using the old template

In this case the app would wait for the relationship to be accepted (and the user could not send a new request) until the template expires (because then the server rejects the pending RelationshipChanges for this template).

One solution to this would be to store a list of "active" relationship_template_ids on the server so we always know if a pending RelationshipChange for a given template is outdated. In the example above this would mean the user would still encounter the same strange behavior as you did but at least the user had to trigger the refresh themself (or wait the 60 mins until the code expires) and they also don't get "locked" because the app is waiting on a lost RelationshipChange.

If we wanted to solve this entirely I think the only way is to store a list of all relationship_template_ids that were created for a given user. This way we can still match "old" templates to their user after a refresh. Theoretically we would have to store these forever because there is no limit on the delay the app could send a RelationshipChange for a template (even if it expired in the meantime). We could limit it to something like the most recent 20 templates per user though. 
With this approach we could also reduce the validity time to something like 60 seconds because even after the template expired we still have the matching to the user. The user could still get "locked" if they scan the code, refresh 20 times and the submit their data but because the code soon expires the user only gets "locked" for these 60 seconds at most. (and this case is rather unlikely)

Reply:
I increased the validity to 10 minutes and we are storing the nbp_uid in RelationshipTemplates now. This should make it much more robust. Maybe we should make the QR code live even longer though. With the nbp_uid stored in the templates now it should be impossible for a RelationshipChange to get "lost" and the user "locked".

Reply:
Thanks, I like that :+1:

With the current implementation, I was giving the system another try. Basically, I made two discoveries:

1. There is a small difference between the displayed validity (in seconds) and the actual QR code validity. This is only a few seconds, but noticeable: Even when the QR code still shows two or three seconds left, I can no longer scan the QR code with the app. We might account for that, by making the QR valid for a longer time span than reporting to the user (like 5 or 10 seconds or so).
2. As soon as the QR code expired, I can no longer scan it with the app (that's fine and expected, a proper error message is shown). However, once I scanned it at a remaining validity of 10 seconds, and then take 20 seconds to provide the required data and submit it, the contact request is just "gone". It neither shows up at "contacts" nor "contact requests" in the app. Further, I was able to verify that the connector doesn't list it either at `/api/v2/Relationships`, so we probably cannot handle it. Given the addition of the `nbp_uid`, do you see any difficulty making the QR code valid for a (much longer) validity to work around this issue? In another experiment, where I scanned the QR code, closed the current browser tab, submitted my details, and re-started the sign-in workflow once again, the relationship was just accepted as expected (even when the validity of the relationship request expired in the mean time).

Reply:
Yes, thats what I meant. Increasing the QR code validity a lot should be fine now. I think I will make it valid for 12 hours. I would keep the expiration time limited though, just in case.

Reply:
Awesome, pretty good! A final remark on this thread: Could you please add a short comment for the validity (i.e., what happens when it expired, why we decided for this 12 hours) and for the `- 5.seconds` in the frontend? I just want to ensure we don't loose something.